text/dependencies/rfcs/0000-dependency-management-phase-one.md

```diff
@@ -0,0 +1,446 @@
+Some dependencies will need to be compiled separately for different stack
+compatibility, and using a Github Action-style set up allows maintainers to
+define their own Dockerfile(s) for the build environments. The workflow that
+will consume this action will be enumerated in the Phase 2 RFC.
+
+Compilation code will go inside of `<buildpack>/dependency/actions/compile`.
+The directory structure will be:
+- `entrypoint` - either a bash script or a directory that contains the compilation code.
+- `<target>.Dockerfile` - one or more Dockerfiles with the target name as a prefix,
+  defining the build environment for compilation. There will be one Dockerfile
+  for each variant of the dependency.
+- `action.yml` - the Github-specific part of the Github Action, which defines
+  the inputs (`version`, `outputDir`, and `target`). It will be a [`composite`
+  type](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action)
+  action, running individual steps to execute compilation on the target
+  Dockerfile. This is done so that the Dockerfile used can be dynamically
+  selected during the workflow execution.
+  
+  An example `action.yml` file for compilation can be seen
+  [here](https://gist.github.com/sophiewigmore/3d09108ddf0fd00f51d183acc8f8f940).
+
+  The compilation action can easily be run locally to produce artifacts by running:
+  ```
+  $ docker build -t compilation -f <target>.Dockerfile <buildpack>/dependency/actions/compile
+  $ docker run -v <output dir>:$PWD compilation --version <version> --outputDir $PWD --target <target>
+  ```
+
+#### Transfer Compilation Code
+
+For any dependencies that must still be compiled or processed in some way, the
+code from the Cloud Foundry repositories should be rewritten, and moved over to
+the entrypoint of the compilation action.
+
+Since the dependency will possibly need to be compiled against different
+stacks/architectures, the different targets should be enumerated in the
+dependency-specific RFC written by buildpack maintainers. This should include
+information about the different images we will need to compile against, as well
+as the stacks that will be compatible with that dependency.
+
+#### Specifiy variants with the `targets.json` file
+A file under the path `<buildpack>/dependency/targets.json` will be an
+enumeration of the different variants of a dependency, based on the different
+stacks the buildpack supports. This will be parsed and used by workflows
+described in Phase 2, but also serves as a manifest of the different version
+variants of each dependency, regardless of whether the dependency is compiled.
+
+The targets listed in the `target.json` will be used during compilation to
+determine which Dockerfile to run on, and will also appear in the name of the
+compiled dependency.
+
+```
+[
+  {
+    "target": <generic name of the group, to be used during compilation>,
+    "stacks": [ <list of compatible stacks>]
+  },
+  ...
+]
+```
```

Comment:
Now that there will be 1 Dockerfile per target inside `dependency/actions/compile` (if I'm understanding that correctly), do you think there's a clever way we could avoid needing a `targets.json` alltogether? Part of me thinks the fewer descriptive files we have to check in, the better. What do others think about this?

Reply:
Yea, I definitely see your point and agree it would be nice to get rid of this, it's unfortunate to include a file just for the sake of making the automation run more smoothly... The `targets.json`  provides a nice way to iterate over the variants of a dependency, and have that laid out clearly. The workflow outlined in Phase 2 heavily depends on iterating over a set of variants that come from each target. 

I think we could get away with removing the `targets.json` file, and relying on the `retrieval` step to be hardcoded to output a batch of metadata for each target variant. Automation could then iteration over each metadata set to perform compilation /  testing. We'd probably want to add a `target` output to the metadata set.

In my mind, the `targets.json` strategy may be clearer because maintainers have to explicitly lay out the separate variants and their compatible stacks. They'd still have to do this in the other proposal I just outlined, but it would be more hidden away as an implementation detail.

Reply:
Open to removing the `targets.json` field, interested to hear what others think about this too.

Reply:
I've been mulling this over and I've actually come around to removing the `targets.json` file in favour of spitting out different metadata entries for each target. The latest commit I added has this outlined. I will update Phase 2 now.